server/aesdsocket.c

```diff
@@ -0,0 +1,379 @@
+
+        // Create file to write packets to
+        FILE* data_file = fopen(TMP_FILE, "a+");
+
+        if (data_file == NULL) {
+            syslog(LOG_ERR, "Error fopen(): %s\n", strerror(errno));
+            close(socket_fd);
+            close(client_fd);
+            if (rx_buffer != NULL) {
+                free(rx_buffer);
+                rx_buffer = NULL;
+            }
+            exit_status = 1;
+            continue;
+        }
+
+        while (1) {
+
+            bytes_to_read = rx_size - total_bytes;
+
+            if (bytes_to_read == 0) {
+                // Increase size of rx_buffer
+                rx_size += READ_SIZE;
+                bytes_to_read = READ_SIZE;
+                rx_buffer = realloc(rx_buffer, rx_size);
+                if (rx_buffer == NULL) {
+                    syslog(LOG_ERR, "Error failed to realloc()\n");
+                    fclose(data_file);
+                    close(socket_fd);
+                    close(client_fd);
+                    break; // send to cleanup
+                }
+                memset(&(rx_buffer[total_bytes]), 0, READ_SIZE);
+            }
+
+            rx_bytes = recv(client_fd, &(rx_buffer[total_bytes]), rx_size - total_bytes, 0);
+
+            if (rx_bytes == 0) {
+                break; // done sending
+            }
+
+            total_bytes += rx_bytes;
+
+            // check buffer for '\n' newline character
+            // If it exists, append to file and send back on client 
+            // check for all '\n' characters, after transmitting
+            // move all data down to zero index of malloc'd packet
+            // USE strchr() to fine \n characters: 
+            // https://man7.org/linux/man-pages/man3/strchr.3.html
+            if (total_bytes < rx_size) {
+                while (1) {
+                    // Check for newline
+                    p_end = strchr(rx_buffer, '\n');
+
+                    if (p_end == NULL) {
+                        break; // no newline found
+                    }
+
+                    // Write packet to file
+                    char* p = rx_buffer;
+                    int packet_len = p_end - p;
+                    while (p <= p_end) {
+                        fprintf(data_file, "%c", *p);
+                        p++;
+                        total_bytes--;
+                    }
+
+                    // Remove saved packet and shift data down to start of buffer.
+                    memset(rx_buffer, 0, packet_len);
+                    memcpy(rx_buffer, p, total_bytes);
+
+                    // Send packet back on client
+                    char line[WRITE_SIZE];
+                    memset(line, 0, WRITE_SIZE);
+                    int tx_bytes = 0;
+                    int tx_bytes_to_send = 0;
+
+                    // Reset file pointer to read from beginning of file
+                    rewind(data_file);
+
+                    while (fgets(line, WRITE_SIZE, data_file) != NULL) {
```

Comment:
@mskogen what happens if more than 1024 bytes are between `\n` characters in this file?